### PR TITLE
Make product variations endpoints aware of parent product id provided

### LINKED ISSUE #41373: WC REST API - Variations endpoints ignores the parent id: /wc/v3/products/<product_id>/variations/<variation_id>
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When making the following REST API call: /`wp-json/wc/v3/products/<product_id>/variations/<id>`, the parent id (product_id) is ignored, and it will retrieve the variation independently if the parent id is correct or if exists.

A similar issue happened when updating the variation. See here: https://github.com/woocommerce/woocommerce/issues/18933

In this case, simply fetching the variation data doesn't create any risk, but it can be misleading, considering that the documentation specifies that product_id should be the parent ID.

### Expected behavior

When sending the following request: `wp-json/wc/v3/products/NOT_EXISTING_PRODUCT_ID/variations/REAL_VARIATION_ID`, it should return a 404 error since the parent does not include that variation ID and the parent does not exist.

### Actual behavior

When making the following request: `wp-json/wc/v3/products/NOT_EXISTING_PRODUCT_ID/variations/REAL_VARIATION_ID`
it returns the data from the "REAL_VARIATION_ID". 

### Steps to reproduce

1. Make this request: GET `wp-json/wc/v3/products/NOT_EXISTING_PRODUCT_ID/variations/REAL_VARIATION_ID`
2. See that the API response with the data from the variation id.

### WordPress Environment

WC 8.3.0
WP 6.4.1

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by jorgeatorres**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

[Most variations endpoints](https://woocommerce.github.io/woocommerce-rest-api-docs/#product-variations) take the form `/products/<product_id>/variations/<id>`, but with the exception of `PUT` (update) requests, `<product_id>` was ignored altogether and we only checked that the variation with ID `<id>` exists.

This PR enforces some additional checks to ensure that that the product referred by `<product_id>` also exists and that `<id>` is actually a variation of said product.
This makes the variations endpoints consistent with `/products/<product_id>` which would obviously check that the product exists, so we no longer have a situation where `/products/1234` would fail but `/products/1234/variations/<existing_variation_id>` would be successful.

⚠️ Changes here could be considered backwards incompatible as it'd no longer be possible to get details for a variation or delete it by using `/products/<random int>/variations/<variation_id>`. Related: p1719503648061859-slack-C0E1AV8T0.

Closes #41373.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Generate and take note of your API consumer key and secret in WC > Settings > Advanced > REST API.
2. Go to WC > Products and create at least 2 variable products with at least a couple variations each. Take note of the product IDs.
3. Configure your favorite REST client to use Basic Auth and use your consumer key and secret as username and password, respectively.
4. Perform a GET request to `<yoursite>/wp-json/wc/v3/products/<product_id>/variations/`, where `<product_id>` is one of the IDs from step 2, and confirm that all the variations for that product are listed. Take note of the ID (`id`) for a couple of those variations.
5. Confirm that `GET`, `PUT` and `DELETE` requests to `/wc/v3/products/<product_id>/variations/<id>` where `<product_id>` is a product ID from step 2 and `<id>` is a variation ID from step 4 all work as [described in the REST API documentation](https://woocommerce.github.io/woocommerce-rest-api-docs/#product-variations).
6. Confirm that `GET`, `PUT` and `DELETE` requests to `/wc/v3/products/<product_id>/variations/<id>` all return 404 when...
   - `<product_id>` is the ID of an existing variable product, and `<id>` is either a non-existent ID or the ID of a variation of a different product.
   - `<product_id>` is a non-existent product ID and `<id>` is an existing variation ID.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-product-variations-v2-controller.php [hunk 1]
```diff
@@ -137,10 +137,48 @@ public function register_routes() {
...
+	/**
+	 * Checks that a variation belongs to the specified parent product.
+	 *
+	 * @param int $variation_id Variation ID.
+	 * @param int $parent_id    Parent product ID to check against.
+	 * @return bool TRUE if variation and parent product exist. FALSE otherwise.
+	 *
+	 * @since 9.2.0
+	 */
+	protected function check_variation_parent( int $variation_id, int $parent_id ) {
```
Comment (barryhughes) [inline plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-product-variations-v2-controller.php]: Optional, but we could also specify the return type:

```suggestion
	protected function check_variation_parent( int $variation_id, int $parent_id ): bool {
```

File: plugins/woocommerce/changelog/fix-41373 [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: update
+
+Enforce product parent checks in variations REST API.
```
Comment (barryhughes) [inline plugins/woocommerce/changelog/fix-41373]: Optional, as this is a bit subjective, but what do you think of:

```suggestion
Ensures the product ID is valid when interacting with product variations via the REST API.
```

(I'm just wondering if this or perhaps something else might help people reading the changelog to get a better sense of what changed—but, again, totally optional.)

**Review by barryhughes — APPROVED** [review]
Thanks, @jorgeatorres—this looks great.

Functions exactly as described: this is a nice improvement and should add extra safety.

I left a couple of minor (and optional) comments, so approving but not merging so you have a chance to see those. However, please also feel free to proceed without them.

**Comment by jorgeatorres** [comment]
Thanks for the review and for the suggestions @barryhughes! I've included those and will be merging now. Thanks again!

**Pull request merged**